### PR TITLE
Fix classification when terrain depth writes are enabled

### LINKED ISSUE #6867: 3D Tiles show through the globe
If a tile's geometric error is sufficiently large to be visible at global scale, it will show through the opposite side of the globe.

This happens due to rendering the depth plane after 3D Tiles, which was done to fix a separate set of bugs - the classification and picking bugs below. It only happens when `globe.depthTestAgainstTerrain` is false so once terrain is on by default none of these problems will happen.

Tileset showing through globe - [Sandcastle code](http://localhost:8080/Apps/Sandcastle/#c=ZVJRi9swDP4roi/NQee0jL3semWjN9jDYGVXdi+B4SZqos2xgu0k9Mb998lJ07WbCY5sfZ/0SXKawsdjQAehQvA5WgTDuvBQc4eQ6xqdhsDATcOegmCoQOAjlIYPqGAvtIbJBg89GQO+4l5iOW7LSmW20w46wl4SPIDFHrboqa3V9+EumefDccs2aLLo5nf3IyeQQY/hljT+3j7uR2fyO7Mgq3XmPcyVSuV70nVj8FEHnd6g013UuDXcFlfmM4Vqh2642Dlu0AUS7Dm5+unZzjP7OogaBKuhQ6pxVFOgDr3SRZGc8VewF+Z6z5NjcV3EZ9QF2XJHIa++aVtislTLBbxZxf2d2HcxTmYnEU7wJxFXk0clQ7LJsbV5ILaXvHBuxET5USLXGBzln5zj2PnVclpqeT+Ce7IF90rwe6qR2/A37iXgTVDHLHsuo0IrRnPp4tl+ohccU11yvMaKlrGeoYezxWztw8ngZnR/oLphF+IAE5ldQJmdDjKAQ5v/kpS595G2TifSuqAOqHjIZv+8m2wGudHei+fYGhOlZLPNOhX8DS0+bWn+1w6d0acIqVabL+OlUmqdyvF/VmA2B+2uIv4B) - when the depth plane is rendered after 3D Tiles.

show-through

Classification bug - [Sandcastle code](http://localhost:8080/Apps/Sandcastle/index.html#c=ZVNdb9owFP0rES8NEjX+jO2OVqsAqUjtVlE2aRpT5QZTrCU2cgwdq/rf5xCgZfjF8b3n3Ht8brxWPlkb/aJ9cplY/ZL0dWVWJfi+jaVn+fbYdzYoY7U/a3+a2qldR1Ywha50OKY1GxlMmmT6mqx8cbHPjpwd68qtfK7B3LvyuoqY0SwlgmTt5K2u3WgBVa6tBktvShPMWldAzWbpruNOQgPMVam9AjFcK05fpzaJa6arYKwKxtmLI3nKh/ilLEkRRhmSAkiGMiwRJrKTnFNOOJQMIIkxpBJ1EgoFpIwByBmmXAjB2p2mh/NGR1dOetxoNTP2+d6EfDF2RZFCQDhnEknG61qZELETBJByigSRlEnCM0I7SYxBVDemiIoMQ0Q4ObTTdjbxylZz58vk4OidCt78oWA0GH6ZjCY/pvbtfUJRngmbOKCdV9uz2Zm5c2rpis2zs7HiLlCvhYme+nyxSY5udt9gb/bZ9MTV7VTH8f5R6LX3apP+PEeAIElJJjHnBFGIcX3TTAoBM8wIlLDeoyVbIJGMSQGjCRjxA1Ci6B3nCDP6DiS1VzK6l+2BPLYhmEooskNBSpjAmEbj4ygPOI5kxhjLpBS/2nuL61WqoL1RxbvFfVc4D8bDAXgxYXFdLBcqjpR9JOWFqiozN/n2b5hslvoD/SQH+sOH0be7RzJ4nIxuh02Zt93kWp1WrwqbQl818c+mXDof6meUAtANulwWUWLVfVrlv3UAeVXVtF53T+rNzDoxs8tp67+nO201OmNmviqKB/NXT1tXvW7EH9EKt/1/v661L9SmhizQ1W0TBAD0uvF4ygrOFU/Kf6j4Dw) - when the depth plane is rendered before 3D Tiles and the depth plane offset **is not** applied.

 classification-bug

Picking bug - [Sandcastle code](http://localhost:8080/Apps/Sandcastle/#c=dVJrb9owFP0rVr40SJlDEufFKNrEkIpURjVYpUmRkEkMterYkW3Sson/PufBSot2P/len3PP8fWtsQQ1JS9EglvAyQuYEkUPJXxsa/ZN3qZTwTWmnMibweeMZ7w2LMI11UfD6uiwzSlREBeF/SfjwMSWMrYVWBajvtAELfGejDILQrc9KrfT3NyLvdiImkiGj7Di+8zqSKeMn950S4O4w7xgHy2vckkIX1U4J7PauOlBdu9P5YQTmGPTRDXdLvpARfScVwf9NddUcHt34N2hwZSm1QD0/rtHw0oo2gDeXt91r2j+/NDf/SObyRTnYiN8cv7neH2sCFwsf65mm8Xycda/+b39kkjc+G0+6DzmgihNOW4ERpcTmWKpzQnzwEZo6KEEwTROvShJUOwA30dpGsM0SlKUIi/1AgegMA09L4Je7A9DP/WSgdNJCEmNwWuJO4ILyvcPVOdPPwRjdgT9JPCSMBjGXpyGSeSAT0MYDk1EcZBEAUJR4jvgCndWMtNaS8zVTsgSjM46C6wlfUVw/m32fT1f/+pXwnKssdJHRiYd+QstKyE1OEhmm/3SpKwYNjNwt4f8mWiYq/bvx+6ZNC5oDWhxm1kfNj2zQM6wUuZmd2BsRX+TzJqMXYN/R2OiHcCyW9sG8uRN7rsihHDsmvSapYVgWywvOv4F) - when the depth plane is rendered before 3D Tiles and the depth plane offset **is** applied.

picking-bug

### PR DESCRIPTION
**Pull request by lilleyse**
The stencil buffer now stores a bit that says whether a pixel is 3D Tiles or not. This allows a classification primitive with `classificationType = ClassificationType.CESIUM_3D_TILE` to apply just to 3D Tiles even if the terrain depth is written. This also lets the depth plane get rendered immediately after the terrain pass which solves #6867, so 3D Tiles no longer show through the opposite side of the globe.

This is going into a staging branch and the follow up PR is going to be enabling textured poylgons and polylines on 3D Tiles, which is a lot easier with this change.

To do:
* [x] Fix unit tests
* [x] CHANGES.md

Next:
* [ ] Set `classificationType` default to `BOTH` everywhere.
* [ ] Optimize `BOTH` - https://github.com/AnalyticalGraphicsInc/cesium/pull/7422#issuecomment-448346515
* [ ] Optimize invert classification - https://github.com/AnalyticalGraphicsInc/cesium/pull/7422#issuecomment-448378983
* [ ] Textured polygons and polylines on 3D Tiles

### CODE REVIEW

**Comment by cesium-concierge** [comment]
Thanks for the pull request @lilleyse!
* :heavy_check_mark: Signed CLA found.

Reviewers, don't forget to make sure that:

- [ ] Cesium Viewer works.
- [ ] Works in 2D/CV.
- [ ] Works (or fails gracefully) in IE11.

---

__I am a bot who helps you make Cesium awesome! [Contributions to my configuration are welcome.](https://github.com/AnalyticalGraphicsInc/cesium/tree/master/.concierge)__

:earth_africa: :earth_americas: :earth_asia:

File: Source/Scene/Cesium3DTileBatchTable.js [hunk 1]
```diff
@@ -1261,30 +1263,42 @@ define([
                 derivedCommands.originalCommand = deriveCommand(command);
                 command.dirty = false;
             }
+            var originalCommand = derivedCommands.originalCommand;
```
Comment (lilleyse) [inline Source/Scene/Cesium3DTileBatchTable.js]: Most of the changes in this function are just cleanup rather than bug fixes.

**Comment by bagnell** [comment]
I see 71 test failures after resetting the branch to the latest commit.

There is an optimization you could make when the classification type is both terrain and 3D Tiles without the invert classification. You could render both terrain and 3D Tiles, render the first two commands for each of the classification types, and then issue a single color draw call. Given that classification primitives are expensive to render, this might be a significant performance improvement, but it doesn't have to hold up this PR.

**Comment by lilleyse** [comment]
I'm still fixing up the unit tests and will bump when they're ready.

Yeah that idea might work. It will be a good win too because in the next PR I might set all the `classificationType` defaults to `BOTH`.

I also have ideas for speeding up inverse classification. The opaque version wouldn't need a separate FBO, just one full screen pass on the main framebuffer. I think the translucent version can maybe be shrunk to a single FBO rather than two, but I haven't thought it through enough yet.

**Comment by lilleyse** [comment]
@bagnell this is ready. All test pass locally now and the Sandcastle demos looks normal. There was a bug with translucent invert classification that is now fixed.

One change I made while writing tests was to set `Vector3DTileGeometry` and `Vector3DTilePolygons` `classificationType` to `ClassificationType.BOTH` by default. I don't think this actually affects anything other than the tests.

**Comment by bagnell** [comment]
This looks good to me. Merging this will close the bugs. Should they stay open until the staging branch is merged?

**Pull request merged**